File: backend/app/services/audit_service.py
from datetime import datetime, timezone
from typing import Optional, Dict, Any, List
from sqlalchemy.orm import Session
from sqlalchemy import desc
from app.models.audit_log import AuditLog

class AuditService:
    @staticmethod
    def create_log(
        db: Session,
        record_type: str,
        asset_id: str,
        hash_value: str,
        version: str = "v1.0",
        contributor_name: str = "System",
        status: str = "VERIFIED",
        is_demo: bool = True,
        metadata_json: Optional[Dict[str, Any]] = None
    ) -> AuditLog:
        """Create an immutable audit trail record."""
        audit = AuditLog(
            record_type=record_type,
            asset_id=asset_id,
            hash=hash_value,
            version=version,
            contributor_name=contributor_name,
            status=status,
            is_demo=is_demo,
            metadata_json=metadata_json,
            timestamp=datetime.now(timezone.utc)
        )
        db.add(audit)
        db.commit()
        db.refresh(audit)
        return audit

    @staticmethod
    def get_logs(
        db: Session,
        skip: int = 0,
        limit: int = 100,
        record_type: Optional[str] = None,
        asset_id: Optional[str] = None
    ) -> List[AuditLog]:
        """Retrieve audit logs with optional filtering and pagination."""
        query = db.query(AuditLog)
        if record_type:
            query = query.filter(AuditLog.record_type == record_type)
        if asset_id:
            query = query.filter(AuditLog.asset_id == asset_id)
        return query.order_by(desc(AuditLog.timestamp)).offset(skip).limit(limit).all()

    @staticmethod
    def get_logs_by_asset(db: Session, asset_id: str) -> List[AuditLog]:
        return db.query(AuditLog).filter(AuditLog.asset_id == asset_id).order_by(desc(AuditLog.timestamp)).all()

audit_service = AuditService()

File: frontend/src/services/auditService.ts
import { apiClient } from './apiClient';
import type { AuditLog } from '../types';

function mapAuditLog(a: any): AuditLog {
  return {
    id: a.id,
    recordType: a.record_type,
    assetId: a.asset_id,
    hash: a.hash,
    version: a.version,
    contributorName: a.contributor_name,
    timestamp: a.timestamp,
    transactionReference: a.transaction_reference,
    status: a.status,
    isDemo: a.is_demo,
  };
}

export const auditService = {
  async getLogs(skip = 0, limit = 100, recordType?: string, assetId?: string): Promise<AuditLog[]> {
    let url = `/audit?skip=${skip}&limit=${limit}`;
    if (recordType) url += `&record_type=${encodeURIComponent(recordType)}`;
    if (assetId) url += `&asset_id=${encodeURIComponent(assetId)}`;

    const res = await apiClient.get<any[]>(url);
    return res.map(mapAuditLog);
  },

  async createLog(log: Omit<AuditLog, 'id' | 'timestamp' | 'transactionReference'>): Promise<AuditLog> {
    const res = await apiClient.post<any>('/audit', {
      record_type: log.recordType,
      asset_id: log.assetId,
      hash: log.hash,
      version: log.version,
      contributor_name: log.contributorName,
      status: log.status,
      is_demo: log.isDemo !== false,
    });
    return mapAuditLog(res);
  },

  async getLogsByAsset(assetId: string): Promise<AuditLog[]> {
    const res = await apiClient.get<any[]>(`/audit/${assetId}`);
    return res.map(mapAuditLog);
  },
};
